### PR TITLE
x265: fix compiling with clang

### LINKED ISSUE #330202: Build failure: pkgsLLVM.x265
### Steps To Reproduce

1. build `pkgsLLVM.x265`

### Build log

```
make[2]: *** [CMakeFiles/x265-shared.dir/build.make:90: ipfilter8.S.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:227: CMakeFiles/x265-shared.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
<instantiation>:2:9: error: unknown directive
        .endfunc
        ^
<instantiation>:32:1: note: while in macro instantiation
endfunc
^
/build/x265_3.5/source/common/aarch64/ipfilter8.S:412:1: note: while in macro instantiation
LUMA_HPS 64 32
^
<instantiation>:11:9: error: unknown directive
        .func x265_interp_8tap_horiz_ps_64x48_neon
        ^
<instantiation>:1:1: note: while in macro instantiation
function x265_interp_8tap_horiz_ps_64x48_neon
^
/build/x265_3.5/source/common/aarch64/ipfilter8.S:413:1: note: while in macro instantiation
LUMA_HPS 64 48
^
<instantiation>:2:9: error: unknown directive
        .endfunc
        ^
<instantiation>:32:1: note: while in macro instantiation
endfunc
^
/build/x265_3.5/source/common/aarch64/ipfilter8.S:413:1: note: while in macro instantiation
LUMA_HPS 64 48
^
<instantiation>:11:9: error: unknown directive
        .func x265_interp_8tap_horiz_ps_64x64_neon
        ^
<instantiation>:1:1: note: while in macro instantiation
function x265_interp_8tap_horiz_ps_64x64_neon
^
/build/x265_3.5/source/common/aarch64/ipfilter8.S:414:1: note: while in macro instantiation
LUMA_HPS 64 64
^
<instantiation>:2:9: error: unknown directive
        .endfunc
        ^
<instantiation>:32:1: note: while in macro instantiation
endfunc
^
/build/x265_3.5/source/common/aarch64/ipfilter8.S:414:1: note: while in macro instantiation
LUMA_HPS 64 64
^
make[2]: *** [CMakeFiles/x265-static.dir/build.make:90: ipfilter8.S.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:147: CMakeFiles/x265-static.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

### Notify maintainers

@codyopel 

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by RossComputerGuy**
## Description of changes

Fixes #330202

Use GCC for assembly because Clang doesn't support the `.endfunc` directive.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by AndersonTorres — APPROVED** [review]
The whole file could be modernized too. 

Besides, LGTM.

**Comment by AndersonTorres** [comment]
Ops! Red eggs.
Cherry this over Staging.

P.S.: before doing this, put the PR in draft mode, or else you will ping the planet.

**Comment by RossComputerGuy** [comment]
@AndersonTorres I already got the rebuild count down. Shouldn't need to go to staging now.

**Review by Atemu — CHANGES_REQUESTED** [review]
I'm getting a bunch of

```
x265> x86_64-unknown-linux-gnu-ld: error: undefined reference due to --no-allow-shlib-undefined: x265_pixel_sad_64x32_avx512                                                                        
x265> >>> referenced by libx265.so.199                                                                                                                                                              
```

and

```
x265> x86_64-unknown-linux-gnu-ld: error: undefined symbol: x265_scanPosLast_x64                                                                                                                    
x265> >>> referenced by asm-primitives.cpp                                                        
x265> >>>               asm-primitives.cpp.o:(x265::setupAssemblyPrimitives(x265::EncoderPrimitives&, int)) in archive ../libx265.a                                                                 
```

errors in the end. Please verify this actually builds.

**Comment by RossComputerGuy** [comment]
> I'm getting a bunch of

I'm not able to reproduce those errors.

File: pkgs/development/libraries/x265/default.nix [hunk 1]
```diff
@@ -66,7 +67,9 @@ stdenv.mkDerivation rec {
     })
     # Fix detection of NEON (and armv6 build) :
     ./fix-neon-detection.patch
-  ];
+  ]
+    # CMake files require a bit of patching
+    ++ lib.optional (stdenv.cc.isClang && !stdenv.targetPlatform.isDarwin) ./fix-clang-asm.patch;
```
[the remark this answers is not in the document]
Comment (RossComputerGuy) [reply pkgs/development/libraries/x265/default.nix:72]: I made it

File: pkgs/development/libraries/x265/default.nix [hunk 2]
```diff
@@ -89,7 +92,9 @@ stdenv.mkDerivation rec {
     (mkFlag ppaSupport "ENABLE_PPA")
     (mkFlag vtuneSupport "ENABLE_VTUNE")
     (mkFlag werrorSupport "WARNINGS_AS_ERRORS")
-  ];
+  ]
+    # Clang does not support the endfunc directive so use GCC.
+    ++ lib.optional (stdenv.cc.isClang && !stdenv.targetPlatform.isDarwin) "-DCMAKE_ASM_COMPILER=${gccStdenv.cc}/bin/${gccStdenv.cc.targetPrefix}gcc";
```
[the remark this answers is not in the document]
Comment (RossComputerGuy) [reply pkgs/development/libraries/x265/default.nix:97]: Yeah, it's necessary

**Comment by RossComputerGuy** [comment]
Should work now for x86_64-linux, it was working for me because I use aarch64-linux.

**Comment by RossComputerGuy** [comment]
@ofborg build pkgsLLVM.x265

File: pkgs/development/libraries/x265/default.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (Atemu) [reply pkgs/development/libraries/x265/default.nix:72]: Please add the info then.

**Review by Atemu — CHANGES_REQUESTED** [review]
```
x265> [ 98%] Built target encoder     
x265> [ 98%] Built target common     
x265> make[2]: *** No rule to make target 'common/CMakeFiles/common.dir/x86/pixel-a.asm.o', needed by 'libx265.a'.  Stop.     
x265> make[1]: *** [CMakeFiles/Makefile2:121: CMakeFiles/x265-static.dir/all] Error 2                                                                                       
x265> make: *** [Makefile:136: all] Error 2     
x265> make: Leaving directory '/build/x265_3.5/source/build-10bits'     
x265> /nix/store/k0x759gbjakj7zkg2cb0g88ksl68713x-stdenv-linux/setup: line 193: pop_var_context: head of shell_variables not a function context     
```

File: pkgs/development/libraries/x265/default.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (RossComputerGuy) [reply pkgs/development/libraries/x265/default.nix:72]: Will do tonight after work

File: pkgs/development/libraries/x265/default.nix [hunk 3]
```diff
@@ -66,7 +67,9 @@ stdenv.mkDerivation rec {
     })
     # Fix detection of NEON (and armv6 build) :
     ./fix-neon-detection.patch
-  ];
+  ]
```
Comment (paparodeo) [inline pkgs/development/libraries/x265/default.nix:70]: alternatively just remove `.endfunc` and `.func`. they have no effect on the assembled code anyway as they are pseudo opts related to symbol generation. one can verify this by assembling with gas adding and removing the ops. 

```diff
diff --git a/pkgs/development/libraries/x265/default.nix b/pkgs/development/libraries/x265/default.nix
index 7f44834d4867..4285aac6e3c3 100644
--- a/pkgs/development/libraries/x265/default.nix
+++ b/pkgs/development/libraries/x265/default.nix
@@ -72,6 +72,10 @@ stdenv.mkDerivation rec {
     substituteInPlace cmake/Version.cmake \
       --replace "unknown" "${version}" \
       --replace "0.0" "${version}"
+
+    for file in common/{aarch64,arm}/asm.S; do
+      sed -i '/\.\(end\)\?func/d' $file
+    done
   ''
   # There is broken and complicated logic when setting X265_LATEST_TAG for
   # mingwW64 builds. This bypasses the logic by setting it at the end of the
```

**Comment by RossComputerGuy** [comment]
Should work now for x86_64-linux, it was working for me because I use aarch64-linux.

> ```
> x265> make[2]: *** No rule to make target 'common/CMakeFiles/common.dir/x86/pixel-a.asm.o', needed by 'libx265.a'.  Stop.     
> x265> make[1]: *** [CMakeFiles/Makefile2:121: CMakeFiles/x265-static.dir/all] Error 2
> ```

I do not understand how this became a problem

File: pkgs/development/libraries/x265/default.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (RossComputerGuy) [reply pkgs/development/libraries/x265/default.nix:72]: Found an upstream issue report https://bitbucket.org/multicoreware/x265_git/issues/550/new-aarch64-asm-code-does-not-build-with

**Comment by RossComputerGuy** [comment]
I managed to simplify the patch and I am working on upstreaming it.

**Comment by RossComputerGuy** [comment]
Ok, x86_64-linux builds now that the patch is simplified.

File: pkgs/development/libraries/x265/default.nix (hunk 3 above)
Comment (paparodeo) [reply pkgs/development/libraries/x265/default.nix:70]: see also: https://android.googlesource.com/platform/ndk/+/master/docs/ClangMigration.md#no

File: pkgs/development/libraries/x265/default.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (RossComputerGuy) [reply pkgs/development/libraries/x265/default.nix]: https://mailman.videolan.org/pipermail/x265-devel/2024-July/013734.html

[the remark this answers is not in the document]
Comment (Atemu) [reply pkgs/development/libraries/x265/default.nix]: Please add the link as a comment.

[the remark this answers is not in the document]
Comment (RossComputerGuy) [reply pkgs/development/libraries/x265/default.nix]: I'm about to :upside_down_face:

[the remark this answers is not in the document]
Comment (RossComputerGuy) [reply pkgs/development/libraries/x265/default.nix]: Done

**Comment by paparodeo** [comment]
> I managed to simplify the patch and I am working on upstreaming it.

seems like just removing the `.endfunc` and `.func` is the way to go given they don't do anything. this PR seems to be over-complicating things but i guess yolo

**Comment by RossComputerGuy** [comment]
Yeah, it is an option but I like the patch route since I already put in the work.

File: pkgs/development/libraries/x265/default.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (Atemu) [reply pkgs/development/libraries/x265/default.nix]: Just chatting in GH comments :D

**Review by Atemu — APPROVED** [review]

**Comment by paparodeo** [comment]
> Yeah, it is an option but I like the patch route since I already put in the work.

i understand not wanting to abandon the work you put in but it went in the wrong direction and now has to be maintained. it is not really a big deal either way but also not good software development. anyway -- i'm checking out, i made my point.

**Pull request merged**